### ISSUE TITLE
Abstract over Conn and Transaction

### ISSUE THREAD

**Issue by roy-ganz** [blackbeam/mysql_async#164]
Hi. Is it possible to abstract over a connection and a transaction?

Situation: I'm working on an ORM layer and want to  support MySql. For that I need to wrap a connection or transaction with a struct. On that struct I call my library functions. 

From the API I see two ways:
1. `Deref<Target=Conn>`, can't make this work, because it's auto implemented by Rust for `&Conn` and in your API for `Transaction`.
-> To go this way it would be needed to implement it for `Conn` [See Playground]( https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=14088e8a673bfeca43880ebbcdd556a5)
2. `ToConnection`, perfect - but it seems to be a marker trait. There is no getter function on it.

What do you suggest?
Thanks for taking your time.

**Comment by blackbeam** [comment]
Hi.
I'm using `ToConnection` for this purpose, and then relying on the [`Query`][1] trait. Example (it's a copy-paste from one of my projects, so you should ignore some details. The point here is that any valid query is first converted into `QueryWithParams<Cow<'static, str>, Params>`, and then executed using `Query::run` and `T: ToConnection`):

```rust
pub struct DbQuery<T> {
    query: QueryWithParams<Cow<'static, str>, Params>,
    extractor: T,
}

impl<T: Extractor> DbQuery<T> {
    pub fn new<U, V>(query: QueryWithParams<U, V>, extractor: T) -> Self
    where
        U: Into<Cow<'static, str>>,
        V: Into<Params>,
    {
        Self {
            query: QueryWithParams {
                query: query.query.into(),
                params: query.params.into(),
            },
            extractor,
        }
    }

    pub async fn run<'a, 't: 'a, C>(self, conn: C) -> crate::Result<T::Output>
    where
        C: ToConnection<'a, 't> + 'a,
    {
        let res = self.query.run(conn).await?;
        self.extractor.extract(res).await
    }
}
```

Also I believe that it is possible to come up with an `enum`-based newtype and then implement `Deref<Target=Conn>` on it (something like the private [`Connection`][2] structure).

[1]: https://docs.rs/mysql_async/0.28.0/mysql_async/prelude/trait.Query.html
[2]: https://docs.rs/mysql_async/0.28.0/src/mysql_async/connection_like/mod.rs.html#15

**Comment by roy-ganz** [comment]
Thank you for your fast response!
Yes, maybe an `enum` would do the trick. Havn' thought of that, I'll check it out.

My problem was, that I couldn't access the affected rows and the last_insert_id.
That's why I tried to get a connection from whatever source (`Conn` or `Transaction`) the user would provide.
Basically `Queryable` abstracts over those sources. A workaround I found is to query the affected_rows and last_insert_id from the db, but I guess getting them from the `Connection` would be faster.

I'm closing the issue, thanks again and have a nice weekend.
